src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -644,116 +663,198 @@ void HNSWIndex<DataType, DistType>::getNeighborsByHeuristic2(
-                candidates.emplace(this->dist_func(getDataByInternalId(neighbor_neighbors[j]),
-                                                   getDataByInternalId(selectedNeighbor),
-                                                   this->dim),
-                                   neighbor_neighbors[j]);
-                neighbors_bitmap[neighbor_neighbors[j]] = true;
-            }
+            neighbor_lock =
+                std::unique_lock<std::mutex>(element_neighbors_locks_[selected_neighbor]);
+            node_lock = std::unique_lock<std::mutex>(element_neighbors_locks_[new_node_id]);
+        }
 
-            idType removed_links[sz_link_list_other + 1];
-            size_t removed_links_num;
-            removeExtraLinks(candidates, Mcurmax, neighbor_neighbors, neighbors_bitmap,
-                             removed_links, &removed_links_num);
-
-            // remove the current neighbor from the incoming list of nodes for the
-            // neighbours that were chosen to remove (if edge wasn't bidirectional)
-            auto *neighbour_incoming_edges = getIncomingEdgesPtr(selectedNeighbor, level);
-            for (size_t i = 0; i < removed_links_num; i++) {
-                idType node_id = removed_links[i];
-                auto *node_incoming_edges = getIncomingEdgesPtr(node_id, level);
-                // if we removed cur_c (the node just inserted), then it points to the current
-                // neighbour, but not vise versa.
-                if (node_id == cur_c) {
-                    neighbour_incoming_edges->push_back(cur_c);
-                    continue;
-                }
+        // get the updated count - this may change between iterations due to releasing the lock.
+        linkListSize cur_node_neighbors_count = getListCount(new_node_neighbors_list);
+        idType *neighbor_neighbors_list = get_linklist_at_level(selected_neighbor, level);
+        linkListSize neighbor_neighbors_count = getListCount(neighbor_neighbors_list);
 
-                // if the node id (the neighbour's neighbour to be removed)
-                // wasn't pointing to the neighbour (i.e., the edge was uni-directional),
-                // we should remove the current neighbor from the node's incoming edges.
-                // otherwise, the edge turned from bidirectional to
-                // uni-directional, so we insert it to the neighbour's
-                // incoming edges set.
-                auto it = std::find(node_incoming_edges->begin(), node_incoming_edges->end(),
-                                    selectedNeighbor);
-                if (it != node_incoming_edges->end()) {
-                    node_incoming_edges->erase(it);
-                } else {
-                    neighbour_incoming_edges->push_back(node_id);
-                }
-            }
+        if (cur_node_neighbors_count == max_M_cur) {
+            // The new node cannot add more neighbors
+            continue;
+        }
```

Comment:
Why `continue` and not `break`?

Reply:
right. Even in the pseudo-code in the doc it says break :)